internal/webhook/external/receiver.go

```diff
@@ -0,0 +1,111 @@
+package external
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+// NewReceiver builds webhook receiver details from the provided base URL,
+// Project name, Secret namespace, and configuration. Project name may be empty
+// for a cluster-level receiver.
+func NewReceiver(
+	ctx context.Context,
+	c client.Client,
+	baseURL string,
+	project string,
+	secretNamespace string,
+	cfg kargoapi.WebhookReceiverConfig,
+) (*kargoapi.WebhookReceiver, error) {
+	receiverType, secretVals, err :=
+		getReceiverTypeAndSecretValues(ctx, c, secretNamespace, cfg)
+	if err != nil {
+		return nil, err
+	}
+	receiver := &kargoapi.WebhookReceiver{
+		Name: cfg.Name,
+		Path: generateWebhookPath(project, cfg.Name, receiverType, secretVals),
+	}
+	if receiver.URL, err = url.JoinPath(baseURL, receiver.Path); err != nil {
+		return nil, fmt.Errorf("error generating webhook URL: %w", err)
+	}
+	return receiver, nil
+}
+
+// getReceiverTypeAndSecretValues determines the type of webhook receiver
+// described by the provided configuration, retrieves necessary secret values
+// for that receiver type and configuration, and returns both.
+func getReceiverTypeAndSecretValues(
+	ctx context.Context,
+	c client.Client,
+	secretNamespace string,
+	cfg kargoapi.WebhookReceiverConfig,
+) (string, []string, error) {
+	var receiverType, secretName string
+	var targetKeys []string
+	switch {
+	case cfg.GitHub != nil:
+		receiverType = kargoapi.WebhookReceiverTypeGitHub
+		secretName = cfg.GitHub.SecretRef.Name
+		targetKeys = []string{kargoapi.WebhookReceiverSecretKeyGithub}
+	default:
+		return "", nil, errors.New(
+			"webhook receiver config has no configuration for a known receiver type",
+		)
```

Comment:
I assume the idea is still that this will be refactored eventually? As I feel this should be put at a higher level.

Reply:
I intended to untangle this further and missed it.